### PR TITLE
Experimental feature: opt-in fast build

### PR DESCRIPTION
**Pull request by sydney-runkle**
This opt in config setting allows for skipping the memory + time heavy parent namespace preparation logic needed for edge cases involving model rebuilds, forward ref simplification, and typevar substitution.

With future refactors, we can hopefully make this speedup the default by eliminating the need for the `__pydantic_parent_namespace__` cache, but that will take significantly more effort.

This is a more practical follow up to https://github.com/pydantic/pydantic/pull/10026.

Closes https://github.com/pydantic/pydantic/issues/8652
Closes https://github.com/pydantic/pydantic/issues/9905

I think for now it's fair to say this closes the following issues, though as mentioned above, there's lots of room for improvement in the default case. I'll open an issue to reflect this.

### CODE REVIEW

**Comment by sydney-runkle** [comment]
Seeing a 5-10x speedup on the examples given in the above issues, specifically `k8_v2.py` models, with this small modification:

```py
from pydantic import BaseModel as PydanticBaseModel, Field, ConfigDict

class BaseModel(PydanticBaseModel):
    model_config = ConfigDict(experimental_fast_build=True)
```

```
python pydanticv2.py  7.17s user 0.21s system 95% cpu 7.693 total
```

Compared to the performance on `main`:

```
python pydanticv2.py  34.71s user 30.23s system 79% cpu 1:21.32 total
```

**Comment by sydney-runkle** [comment]
Happy to get some feedback on the naming of this experimental config flag.

**Comment by sydney-runkle** [comment]
@MarkusSintonen, curious to hear your feedback as well 👍

File: pydantic/config.py [hunk 1]
```diff
@@ -1014,6 +1014,17 @@ class Model(BaseModel):
...
 
+    experimental_fast_build: bool
+    """Build models faster by skipping the logic associated with storing the parent namespace on the model.
+
+    Defaults to `False` to maintain backwards compatibility.
+
+    !!! warning
+        If this setting is set to `True`, model schemas for generic typed dicts and named tuples might be incorrect.
+        There may also be other edge cases where the model schema is incorrect. That being said, this setting has the potential
+        to speed up schema builds by a factor of 10x in cases with hundreds or thousands of complex models.
```
Comment (samuelcolvin) [inline pydantic/config.py:1025]: I'd suggest anyone having issues reports this. 

The ideal case is that this change or it's spiritual successor can become the default - I think we should say that explicitly.

**Review by samuelcolvin — APPROVED** [review]
Lgtm

File: pydantic/config.py [hunk 2]
```diff
@@ -1014,6 +1014,23 @@ class Model(BaseModel):
...
 
+    experimental_fast_build: bool
+    """Build model schemas faster by skipping the logic associated with storing the parent namespace on the model.
+
+    Defaults to `False` to maintain backwards compatibility, for now. The ideal case is that this change or it's successor
+    can become the default. Thus, if you experience issues with this setting that have not already been documented,
+    please open an issue on GitHub.
+
+    !!! warning
+        If this setting is set to `True`, model schemas for models that involve edge case uses of namespaces, such as
```
Comment (MarkusSintonen) [inline pydantic/config.py:1025]: Could we add unit tests showing which cases have issues when `experimental_fast_build=True`? So asserting the exception where they are given in the problem cases. Then its easier to understand what is the impact of using it. Docs could link to the unit test file showing the issues even.

Comment (samuelcolvin) [reply pydantic/config.py:1025]: Great idea.

Comment (MarkusSintonen) [reply pydantic/config.py:1025]: Also would be nice to run some tests with `experimental_fast_build=True` where its expected to work.

**Comment by sydney-runkle** [comment]
Not sure why those tests are failing on CI - they're passing for me locally...

File: pydantic/config.py (hunk 2 above)
Comment (sydney-runkle) [reply pydantic/config.py:1025]: 👍 tests added, thanks for the suggestion!

**Comment by sydney-runkle** [comment]
Alright, I'm using `xfail` with `strict=False` on some of the tests - seems ok, given that we're merely showcasing patterns that might not work with a given setting.

**Comment by Viicos** [comment]
> Alright, I'm using `xfail` with `strict=False` on some of the tests - seems ok, given that we're merely showcasing patterns that might not work with a given setting.

Seems like `test_recursive_generic_typeddict_in_function_1` and 2 were unexpectedly passing, so they could be removed. But yes, let's keep them as we are unsure if they really work.

**Review by Viicos — APPROVED** [review]
Any benefit/drawback of using an env variable instead? I may remember we talked about this but unsure.

**Comment by sydney-runkle** [comment]
> Any benefit/drawback of using an env variable instead? I may remember we talked about this but unsure.

In general, we want to try avoiding mutating global behavior like this bc it might effect models used by third party libraries, which is unpredictable and hard to debug. @adriangb has some strong opinions on this subject, if I recall...

**Comment by sydney-runkle** [comment]
Following up further - we thought about introducing an environment variable for things like.

Also, we already have enough trouble juggling config settings, field constraints, and runtime settings to validation / serialization functions, so I'm inclined to avoid sourcing user preferences from another location.

**Comment by sydney-runkle** [comment]
See https://github.com/pydantic/pydantic/issues/10074 regarding next steps and a by-default plan for this logic.
